contracts/token/ERC20/ERC20Permit.sol

```diff
@@ -0,0 +1,109 @@
+pragma solidity ^0.6.0;
+
+import "./ERC20.sol";
+import "./IERC2612Permit.sol";
+import "../../cryptography/ECDSA.sol";
+import "../../utils/Counters.sol";
+
+/**
+ * @dev Extension of {ERC20} that allows token holders to use their tokens
+ * without sending any transactions by setting {IERC20-allowance} with a
+ * signature using the {permit} method, and then spend them via
+ * {IERC20-transferFrom}.
+ *
+ * The {permit} signature mechanism conforms to the {IERC2612Permit} interface.
+ */
+abstract contract ERC20Permit is ERC20, IERC2612Permit {
+    using Counters for Counters.Counter;
+
+    mapping (address => Counters.Counter) private _nonces;
+
+    bytes32 private immutable _PERMIT_TYPEHASH = keccak256("Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)");
+
+    // Mapping of ChainID to domain separators. This is a very gas efficient way
+    // to not recalculate the domain separator on every call, while still
+    // automatically detecting ChainID changes.
+    mapping (uint256 => bytes32) private _domainSeparators;
+
+    constructor() internal {
+        _updateDomainSeparator();
+    }
+
+    /**
+     * @dev See {IERC2612Permit-permit}.
+     *
+     * If https://eips.ethereum.org/EIPS/eip-1344[ChainID] ever changes, the
+     * EIP712 Domain Separator is automatically recalculated.
+     */
+    function permit(address owner, address spender, uint256 amount, uint256 deadline, uint8 v, bytes32 r, bytes32 s) public virtual override {
+        require(block.timestamp <= deadline, "ERC20Permit: expired deadline");
+
+        bytes32 hashStruct = keccak256(
+            abi.encode(
+                _PERMIT_TYPEHASH,
+                owner,
+                spender,
+                amount,
+                _nonces[owner].current(),
+                deadline
+            )
+        );
+
+        bytes32 hash = keccak256(
+            abi.encodePacked(
+                uint16(0x1901),
+                _domainSeparator(),
+                hashStruct
+            )
+        );
+
+        address signer = ECDSA.recover(hash, v, r, s);
+        require(signer == owner, "ERC20Permit: invalid signature");
+
+        _nonces[owner].increment();
+        _approve(owner, spender, amount);
+    }
+
+    /**
+     * @dev See {IERC2612Permit-nonces}.
+     */
+    function nonces(address owner) public view override returns (uint256) {
+        return _nonces[owner].current();
+    }
+
+    function _updateDomainSeparator() private returns (bytes32) {
+        uint256 chainID = _chainID();
+
+        bytes32 newDomainSeparator = keccak256(
+            abi.encode(
+                keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"),
+                keccak256(bytes(name())),
+                keccak256(bytes("1")), // Version
```

Comment:
It seems we should use pure internal method `version()` or `permitVersion()` for this.

Reply:
Are you suggesting an internal setter for users to set their own version field? I'm actually not sure what all the use cases for `version` are, if you have insights on this that'd be extremely helpful.

That said we do need to provide a way for either the version or domain separator to be queried, as noted [here](version).

Reply:
I believe version makes sense when behaviour changes without address change (upgradablity?). So I mean version method could be overrided by user, or maybe should be even abstract.

Reply:
I see. It'd be great to have some examples to show in the docs.

Regarding implementation, we could have a setter and a getter, where the setter also calls `_updateDomainSeparator` and not increase gas costs of `permit` that way.

Reply:
@k06a In this case though the semantics of the signature are defined by the EIP, so I don't see a reason for `version` to change.